### DISCUSSION TITLE
Choice of subparser groups

### DISCUSSION THREAD

**Discussion by lebrice** [lebrice/SimpleParsing#134] [Ideas]
This is a draft of how I plan to implement a new feature: choice of subgroups of arguments.

This roughly follows what I described previously in https://github.com/lebrice/SimpleParsing/issues/130

Idea:
```python
#inside ArgumentParser.parse_args method:

self.subgroups: list[str] = []
# add all arguments, resolve conflicts, etc.
self.preprocessing() 

if subgroups:
    # do a "first pass" of argument parsing
    # to detect what the subgroup choice was (e.g. "--optimizer adam")
    self.add_help = False
    args, _ = self.parse_known_args(argv)
    
    # parser for the second round:
    parser = ArgumentParser(parents=[self], add_help=True)
    for dest in self.subgroups:
        # add arguments for the Adam optimizer
        # (since it was chosen in the first phase)
        chosen_dataclass_type = getattr(args, dest)
        parser.add_arguments(chosen_dataclass_type, dest=dest, prefix=f"{dest}.")
    # "recurse", which should enable arbitrary nesting
    return parser.parse_args(argv)
else:
    # do what we usually do. No change.
    ...
```

**Comment by idoby** [comment]
This could be nice for unions but I still need to think about implementing polymorphic types.
We also need to support this if the user calls parse_known_args instead of parse_args, to maintain consistency.

**Reply by lebrice** [reply]
Given that `parse_args` just calls `parse_known_args` anyway, you're right, this would be in `parse_known_args` and work for both.

**Reply by lebrice** [reply]
Also, what precisely do you mean by "polymorphic types"?

**Reply by idoby** [reply]
> Given that `parse_args` just calls `parse_known_args` anyway, you're right, this would be in `parse_known_args` and work for both.

Yeah, parse_args would have to check for any remaining args and raise the appropriate error on the command line like argparse does

> Also, what precisely do you mean by "polymorphic types"?

I mean something like:

```python
@dataclass
class OptimizerConfig:
    pass


@dataclass
class SGDConfig(OptimizerConfig):
    momentum: float


@dataclass
class AdamConfig(OptimizerConfig):
    betas: Tuple[float, ...]
```

The advantages to doing this are:
1. I don't have to repeat myself by creating a class AND adding it to a union
2. I don't have to carry the union type everywhere in the program (e.g. `Union[SGDConfig, AdamConfig]` and even worse when I have more than two)
3. The IDE/mypy/etc can check that I've accounted for every possible subtype of OptimizerConfig, e.g.
```python
match optimizer_config:
    case SDGConfig(momentum):
        do_something_with(momentum)
    case AdamConfig(betas):
        ...
    # No default branch is needed
```

_However_, this requires SP to become opinionated about how polymorphic configs should be written on the command line. I'm not sure SP should go there.

**Reply by lebrice** [reply]
Ah I see.

This will be covered by the subgroups feature. The annotation can just be `OptimizerConfig`. No need for unions.


There's also another feature I'd like to introduce more properly soon, that will make it possible to directly create the arguments for the different optimizers, without having to statically create a dataclass.
This new feature is called `Partial`:

```python
from dataclasses import dataclass

from simple_parsing import ArgumentParser
from simple_parsing.helpers.partial import Partial

from torch.optim import Adam, SGD

AdamConfig = Partial[Adam]
SGDConfig = Partial[SGD]
```

These Partial objects are dataclass types, and are partially configured callables. The attributes of these dataclass are the arguments of the callable type or function. When invoked, they pass their attributes as arguments to the wrapped callables. (the arguments can also be passed to the callable, and will overwrite those stored in the `Partial`).

```python

def some_function(v1: int = 123, v2: int = 456):
    """Gives back the mean of two numbers."""
    return (v1 + v2) / 2


@dataclass
class ModelConfig:

    n_layers: int = 10

    sgd_config: SGDConfig = SGDConfig(lr=1e-3)

   # This is equivalent to doing:
    sgd_config: Partial[SGD] = Partial[SGD](lr=1e-3)


    # Combining this with the subgroups feature, we could choose between optimizers:
    optimizer: Partial[Optimizer] = subgroups({
        "adam": Partial[Adam],
        "sgd": Partial[SGD],
    }, default = "adam")
    
    # This can even configure functions!
    some_function_config: Partial[some_function]
```

I'll make another post to give more details on this soon. But basic idea is that indexing the generic `Partial` class creates a dataclass dynamically from the signature of the type or callable that is used as the index.

Then, you can get an optimizer by calling the Partial object, passing it any required arguments:
```python
model_config: ModelConfig = (...)
optimizer_config = model_config.optimizer(model.parameters())
```

Let me know what you think!

**Reply by idoby** [reply]
Interesting idea. It's going to be difficult to generate these classes automatically though. This example builds a lot of knowledge about torch and its optimizers into the library. Perhaps it should be in a separate SP<->torch integration library (or even better - in my library 😆)

**Reply by janvainer** [reply]
@lebrice Cool idea with the `Partial` functionality. I have something similar implemented - it autgenerates and registers a dataclass config for the wrapped class.__init__ or function. It works, but there are several issues to watch out for:

1. Classes generated at runtime are not naively picklable - this becomes an issue for multi-GPU training - you'll have to overwrite the `__reduce__` function for the `Partial` class
2. The issue 1. will probably require you to somehow register the autogenerated configs so that the `__reduce__` function is able to re-instantiate them when unpickling

I would maybe avoid going there with the `Partial` functionality, it can work nicely in specific projects but could get quite messy in general cases.

Edit: (Mistook repr for reduce)

**Reply by lebrice** [reply]
Hey @janvainer thanks for the reply 

Yeah this is something I'll watch out for. However the class construction process is simple and deterministic, so I seeno reason why the pickling behaviour would have to be complicated: simply pickle the target, the Partial class, and then call Partial[target] on the receiving side when unpickling.

Im using a metaclass (PartialMeta) for the Partial class, which all the dynamic config classes inherit from. I think this way, I can have fine-grained control of how the class objects are pickled by overwriting some method like `__deepcopy__` or `__reduce__` or something similar (havent checked the docs in a while).

I"ll make this work in these kind of multi-GPU scenarios before it gets merged.

**Reply by lebrice** [reply]
@idoby this is in no way tied to PyTorch: you can pass any class or callable to Partial.

The example I was showing above would be in user code.
However I"m not against doing something like `simple_parsing.helpers.torch` or something similar.

**Reply by idoby** [reply]
Creating the optimizer directly from the config is torch-specific. Can't do it without knowing the optimizer tajes the model parameters. It should probably be a separate module, yeah.

**Reply by lebrice** [reply]
Not to be annoying, but I believe you are incorrect: This creates a Partial object that behaves just like a `functools.partial`: It just populates some of the arguments of any callable. In this particular case, the Optimizers need a `params` required argument, which can't be parsed from the command-line. Therefore, you still have to pass the model parameters to the `AdamConfig` object to get an `Adam` instance:

```python
# Create a config dataclass that partially configures the Adam optimizer
# (with a --lr, betas, etc arguments)
AdamConfig = Partial[Adam]
# Create a config dataclass that partially configures the Adam optimizer
# (with a different default value for the `lr` argument)
AdamConfig = Partial[Adam](lr=3e-4)

model = nn.Sequential(...)

adam = AdamConfig(model.parameters()) # Use the lr value from the config
adam = AdamConfig(model.parameters(), lr=1.23) # overwrite the value
adam = AdamConfig(model.parameters(), betas=[0.9, 0.99])  # use the lr from the config, overwrite the defaults for betas
```


Is this clearer? Like, this works for any callable, even if it has required positional arguments (like the optimizer classes do).

**Reply by idoby** [reply]
How do you decide which argument can come from the command line and which doesn't? Seems to me like this might work for this particular case but it doesn't generalize so well...

**Reply by lebrice** [reply]
Good question! I'm open to suggestions on this, but at the moment, only the arguments that have an annotation that is understood by simple-parsing will have their arguments added.

It's also possible to do `AdamConfig = config_dataclass_for(Adam, ignore_args=("params",))` to explicitly specify the arguments that should be ignored.

**Comment by janvainer** [comment]
The subgroup parser looks pretty nice and relatively simple :) Can't wait to start using this instead of my hacky code :D

**Comment by lebrice** [comment]
There is now a PR for this feature: #136
